# TRACE
HMM for DNase footprinting and motif matching

# System Requirements

## Hardware Requirements

## Software Requirements


# Installation
Clone a copy of the TRACE repository

```
git clone https://github.com/Boyle-Lab/TRACE.git
```
To make sure that correct version of Python are used and all required packages are installed, we recommend using Conda and creating the environment from the environment.yml file:

```
conda env create -f environment.yml
```
Build TRACE:

```
make all
```

# Usage information
To call TFBSs, TRACE requies a file of regions of interest, files of sequence infomation, read counts, and slopes at each position and a file containing intial model. 
To generate data files in required format, you can use our python script dataProcessing.py. 

Required input:      
<peak_3.file>: A file containing regions of interest. The 3 columns are chromosome number, start position and end position of regions of interest.   
<bam.file>: A bam file of aligned reads from DNase-seq or ATAC-seq.   
<genome.size>: A genome file defining the length of each chromosome.   
<fasta.file>: A sequence file in FASTA format.    
Output:   
<seq.file>: A file containing sequence information of regions from <peak_3.file>, with required format. (see ./data/E2F1_seq.txt).   
<count.file>: A file contains processed read counts at each position in regions from <peak_3.file>.   
<slope.file>: A file contains processed deritives at each position in regions from <peak_3.file>.   

To generat data files using our python script:
```
./src/dataProcessing.py <peak_3.file> <bam.file> <genome.size> <fasta.file> 
```
The default setting will use DNase-seq based prototal. To use ATAC-seq data instead, include ```--ATAC-seq``` argument and choose from pe (pair-end) and se (single-end). If you have preferred output directory and name, set argument ```--prefix```.

To build an intial HMM model:

```
./src/init_hmm.py <TF>
```
It will generate the initial HMM model <init.model.file> for TF of your inteset.  the default setting will generate a 10-motif model, to change the number of extra motifs, set argument ```--motif-number```.   

Besides  <seq.file>, <count.file>, <slope.file> and <init.model.file>,  the main program also requires <peak_7.file>, a file containing regions of interest. First 3 columns and next 3 columns are chromosome number, start position and end position of regions of interest and motif sites, the last column is number of bases overlapping between motif sites and peaks, which can be easily generated by bedtools intersect. 
To perform footprinting: 

```
./esthmm -Q <seq.file> -L <slope.file> -C <counts.file> -I <init.model.file> -O <final.model.file> -P <peak_7.file> -A <output1.file> -B <output2.file> -T <thread.num>
```

It will generate final model <final.model.file>, and a file containing all motif sites included in <peak_7.file> and their marginal posterior probabilities of being active binding site and inactive binding sites.

If a TRACE model is already trianed and you only want to call binding sites based on exsiting model, you can run viterbi step directly:

```
./viterbi -Q <seq.file> -L <slope.file> -C <counts.file> -I <init.model.file> -O <final.model.file> -P <peak_7.file> -A <output1.file> -B <output2.file> -T <thread.num>
```

## Demo
The data folder contains example data for DNase-seq on K562 cell to predict binding sites of E2F1.  For simplicity, we randomly selected 500 DNase-seq peaks in chr1. 

```
./esthmm -Q ./data/E2F1_seq.txt -L ./data/E2F1_slope.txt -C ./data/E2F1_counts.txt -I ./data/init_hmm.txt -O ./data/E2F1_hmm.txt -P ./data/E2F1_peak_7.bed -A ./data/output1.bed -B ./data/output2.bed -T 20
```
